include/irr/builtin/material_compiler/glsl/rasterization/impl.glsl

```diff
@@ -3,150 +3,189 @@
 
 #include <irr/builtin/material_compiler/glsl/common.glsl>
 
-void instr_eval_execute(in instr_t instr, in vec3 L)
+void instr_eval_execute(in instr_t instr, in MC_precomputed_t precomp, inout irr_glsl_LightSample s, inout irr_glsl_AnisotropicMicrofacetCache _microfacet)
 {
 	uint op = instr_getOpcode(instr);
 
 	//speculative execution
 	bsdf_data_t bsdf_data = fetchBSDFDataForInstr(instr);
+	const mat2x3 ior = bsdf_data_decodeIoR(bsdf_data, op);
+	const float ior_scalar = dot(CIE_XYZ_Luma_Y_coeffs, ior[0]);
+	const mat2x3 ior2 = matrixCompMult(ior, ior);
 	params_t params = instr_getParameters(instr, bsdf_data);
 	float bxdf_eval_scalar_part;
 	uint ndf = instr_getNDF(instr);
 	float a = params_getAlpha(params);
 	float a2 = a*a;
-#ifndef ALL_ISOTROPIC_BXDFS
 	float ay = params_getAlphaV(params);
 	float ay2 = ay*ay;
-#endif
+	const vec3 refl = params_getReflectance(params);
 
-	float cosFactor = op_isBSDF(op) ? abs(currBSDFParams.isotropic.NdotL):max(currBSDFParams.isotropic.NdotL,0.0);
+	const bool is_bsdf = !op_isBRDF(op); //note it actually tells if op is BSDF or BUMPMAP or SET_GEOM_NORMAL (divergence reasons)
 
-	if (cosFactor>FLT_MIN && op_hasSpecular(op))
-	{
-#ifdef NDF_GGX
-#ifndef ONLY_ONE_NDF
-		if (ndf==NDF_GGX) {
+#ifndef NO_TWOSIDED
+	handleTwosided(instr, s, _microfacet);
 #endif
 
-#ifdef ALL_ISOTROPIC_BXDFS
-			bxdf_eval_scalar_part = irr_glsl_ggx_height_correlated_cos_eval_DG(currBSDFParams.isotropic, currInteraction.isotropic, a2);
-#else
-			bxdf_eval_scalar_part = irr_glsl_ggx_height_correlated_aniso_cos_eval_DG(currBSDFParams, currInteraction, a, ay);
-#endif
+	const float cosFactor = irr_glsl_conditionalAbsOrMax(is_bsdf, s.NdotL, 0.0);
 
+	uvec3 regs = instr_decodeRegisters(instr);
 
-#ifndef ONLY_ONE_NDF
-		} else
-#endif
+	irr_glsl_AnisotropicMicrofacetCache microfacet;
+	bool is_valid = true;
+	bool refraction = false;
+#ifndef NO_BSDF
+	//here actually using stronger check for BSDF because it's probably worth it
+	if (op_isBSDF(op) && irr_glsl_isTransmissionPath(currInteraction.isotropic.NdotV, s.NdotL))
+	{
+		float orientedEta, rcpOrientedEta;
+		irr_glsl_getOrientedEtas(orientedEta, rcpOrientedEta, currInteraction.isotropic.NdotV, ior[0].x);
+		is_valid = irr_glsl_calcAnisotropicMicrofacetCache(microfacet, true, currInteraction.isotropic.V.dir, s.L, currInteraction.T, currInteraction.B, currInteraction.isotropic.N, s.NdotL, s.VdotL, orientedEta, rcpOrientedEta);
+		refraction = true;
+	}
+	else
 #endif
+	{
+		microfacet = _microfacet;
+	}
 
-#ifdef NDF_BECKMANN
-#ifndef ONLY_ONE_NDF
-		if (ndf==NDF_BECKMANN) {
+	bxdf_eval_t bxdf_eval = bxdf_eval_t(0.0);
+	if (is_valid && cosFactor>FLT_MIN)
+	{
+#if defined(OP_DIFFUSE) || defined(OP_DIFFTRANS)
+		if (op_isDiffuse(op))
+		{
+			vec3 reflectance = is_bsdf ? vec3(1.0) : refl;
+			float alpha2 = is_bsdf ? 0.0 : a2;
```

Comment:
still, whats with the reflectance and roughness modification?

Reply:
this vec3(1.0) is wrong, changed it to transmittance. alpha2==0.0 in case of bsdf is because OP_DIFFTRANS is always smooth (doesnt even have roughness parameter)

Reply:
> because OP_DIFFTRANS is always smooth (doesnt even have roughness parameter)
we could allow a rough DIFFTRANS (in the IR) at no extra cost (excution or complexity)